client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java

```diff
@@ -220,18 +264,46 @@ private CloseableHttpAsyncClient createHttpClient() {
                 //default settings for connection pooling may be too constraining
                 .setMaxConnPerRoute(DEFAULT_MAX_CONN_PER_ROUTE).setMaxConnTotal(DEFAULT_MAX_CONN_TOTAL)
                 .setSSLContext(SSLContext.getDefault())
+                .setUserAgent(String.format(Locale.ROOT, "elasticsearch-java/%s (Java/%s)",
+                    VERSION.isEmpty() ? "Unknown" : VERSION, System.getProperty("java.version"))
+                )
                 .setTargetAuthenticationStrategy(new PersistentCredentialsAuthenticationStrategy());
             if (httpClientConfigCallback != null) {
                 httpClientBuilder = httpClientConfigCallback.customizeHttpClient(httpClientBuilder);
             }
 
+            // Always add metadata header last so that it's not overwritten
+            final String metadataHeaderValue = getMetadataHeaderValue();
+            httpClientBuilder.addInterceptorLast((HttpRequest request, HttpContext context) -> {
+                if (metadataHeaderValue != null) {
+                    request.setHeader(METADATA_HEADER, metadataHeaderValue);
+                } else {
+                    request.removeHeaders(METADATA_HEADER);
+                }
+            });
             final HttpAsyncClientBuilder finalBuilder = httpClientBuilder;
             return AccessController.doPrivileged((PrivilegedAction<CloseableHttpAsyncClient>) finalBuilder::build);
         } catch (NoSuchAlgorithmException e) {
             throw new IllegalStateException("could not create the default ssl context", e);
         }
     }
 
+    private String getMetadataHeaderValue() {
+        if (metadataHeaderEnabled) {
+            VersionInfo hcVersion = VersionInfo.loadVersionInfo("org.apache.http.nio.client",
+                HttpAsyncClientBuilder.class.getClassLoader());
+
+            // service, language, transport, followed by additional information
+            return "es=" + VERSION +
```

Comment:
These are all things that will not change for the lifetime of the jvm, yet http client version and the "runtime metadata" seem to be inspected on every invocation of this method.

Instead, we could have a single method that gets invoked to generate this data, and statically cache the result. We can then be more stringent in failures, since if metadata is enabled, we shouldn't have silent failures, which in that case may only cause confusion. The error handling can be simpler then, with a single try/catch that rethrows as a runtime exception.